src/widget/text_edit.rs

```diff
@@ -575,6 +645,38 @@ impl<'a> Widget for TextEdit<'a> {
                             }
                         },
 
+                        input::Key::End => { // move cursor to end.
+                            let mut line_infos = state.line_infos.iter().cloned();
+                            let line = match cursor {
+                                Cursor::Idx(idx) => idx.line,
+                                Cursor::Selection {end, ..} => end.line, // use last line of selection
+                            };
+                            match line_infos.nth(line) {
+                                Some(line_info) => {
+                                    let char = line_info.end_char() - line_info.start_char;
+                                    let new_cursor_idx = text::cursor::Index { line: line, char: char };
+                                    cursor = Cursor::Idx(new_cursor_idx);
+                                },
+                                _ => (),
+                            }
+                        },
+
+                        input::Key::Home => { // move cursor to beginning.
+                            let mut line_infos = state.line_infos.iter().cloned();
+                            let line = match cursor {
+                                Cursor::Idx(idx) => idx.line,
+                                Cursor::Selection {start, ..} => start.line, // use first line of selection
+                            };
+                            match line_infos.nth(line) {
```

Comment:
This match expr could be more concisely expressed as `if line_infos.nth(line).is_some() { ... }`